File: python/iforevents/integrations/posthog.py
"""PostHog adapter (``pip install "iforevents[posthog]"``)."""
from __future__ import annotations

from typing import Any, Optional

from ..events import IdentifyEvent, PageEvent, TrackEvent
from ..integration import Integration
from ._vendor import load_vendor


class PostHogIntegration(Integration):
    def __init__(self, api_key: Optional[str] = None, *, host: str = "https://us.i.posthog.com", client: Any = None, anonymous_id: str = "server", **hooks: Any) -> None:
        super().__init__(name="PostHogIntegration", **hooks)
        self.api_key = api_key
        self.host = host
        self.client = client
        self.anonymous_id = anonymous_id
        self._distinct_id: Optional[str] = None

    def init(self) -> None:
        super().init()
        if self.client is None:
            if not self.api_key:
                raise ValueError("PostHogIntegration needs an api_key or a client")
            self.client = load_vendor("posthog", "posthog").Posthog(self.api_key, host=self.host)

    def _who(self) -> str:
        return self._distinct_id or self.anonymous_id

    def identify(self, event: IdentifyEvent) -> None:
        super().identify(event)
        self._distinct_id = event.custom_id
        self.client.identify(distinct_id=event.custom_id, properties=event.traits)

    def track(self, event: TrackEvent) -> None:
        super().track(event)
        self.client.capture(distinct_id=self._who(), event=event.name, properties=event.properties, timestamp=event.timestamp)

    def page(self, event: PageEvent) -> None:
        super().page(event)
        props = dict(event.properties, screen_name=event.name, navigation_type=event.navigation_type, to_route=event.to_route, previous_route=event.previous_route)
        self.client.capture(distinct_id=self._who(), event="$pageview", properties=props, timestamp=event.timestamp)

    def reset(self) -> None:
        super().reset()
        self._distinct_id = None

    def flush(self) -> None:
        self.client.flush()

    def shutdown(self) -> None:
        self.client.shutdown()

File: go/integrations/posthog/posthog.go
// Package posthog forwards IForevents calls to PostHog through
// github.com/posthog/posthog-go.
package posthog

import (
	"context"
	"fmt"
	"sync"

	iforevents "github.com/innovafour/iforevents-sdks/go"
	ph "github.com/posthog/posthog-go"
)

// Client is what the adapter needs; *posthog.Client satisfies it.
type Client interface {
	Enqueue(ph.Message) error
	Close() error
}

// Config of the adapter. APIKey creates a client; Client reuses one.
type Config struct {
	APIKey string
	// Endpoint of the PostHog instance. Default https://us.i.posthog.com.
	Endpoint string
	// Configure mutates the posthog.Config before the client is created.
	Configure func(*ph.Config)
	Client    Client
	// AnonymousID is the distinct id used before Identify. Default "server".
	AnonymousID string
	Hooks       iforevents.Hooks
}

type Integration struct {
	iforevents.BaseIntegration
	cfg        Config
	client     Client
	mu         sync.Mutex
	distinctID string
}

func New(cfg Config) *Integration {
	if cfg.AnonymousID == "" {
		cfg.AnonymousID = "server"
	}
	if cfg.Endpoint == "" {
		cfg.Endpoint = "https://us.i.posthog.com"
	}
	return &Integration{BaseIntegration: iforevents.BaseIntegration{IntegrationName: "PostHogIntegration", Hooks: cfg.Hooks}, cfg: cfg, client: cfg.Client}
}

func (p *Integration) Init(ctx context.Context) error {
	if err := p.BaseIntegration.Init(ctx); err != nil {
		return err
	}
	if p.client == nil {
		if p.cfg.APIKey == "" {
			return fmt.Errorf("posthog: Config.APIKey or Config.Client is required")
		}
		c := ph.Config{Endpoint: p.cfg.Endpoint}
		if p.cfg.Configure != nil {
			p.cfg.Configure(&c)
		}
		client, err := ph.NewWithConfig(p.cfg.APIKey, c)
		if err != nil {
			return err
		}
		p.client = client
	}
	return nil
}

func (p *Integration) who() string {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.distinctID != "" {
		return p.distinctID
	}
	return p.cfg.AnonymousID
}

func (p *Integration) Identify(ctx context.Context, e iforevents.IdentifyEvent) error {
	if err := p.BaseIntegration.Identify(ctx, e); err != nil {
		return err
	}
	p.mu.Lock()
	p.distinctID = e.CustomID
	p.mu.Unlock()
	return p.client.Enqueue(ph.Identify{DistinctId: e.CustomID, Properties: ph.Properties(e.Traits)})
}

func (p *Integration) Track(ctx context.Context, e iforevents.TrackEvent) error {
	if err := p.BaseIntegration.Track(ctx, e); err != nil {
		return err
	}
	return p.client.Enqueue(ph.Capture{DistinctId: p.who(), Event: e.Name, Properties: ph.Properties(e.Properties), Timestamp: e.Timestamp})
}

func (p *Integration) Page(ctx context.Context, e iforevents.PageEvent) error {
	if err := p.BaseIntegration.Page(ctx, e); err != nil {
		return err
	}
	props := ph.Properties{}
	for k, v := range e.Properties {
		props[k] = v
	}
	props["screen_name"], props["navigation_type"], props["to_route"], props["previous_route"] = e.Name, e.NavigationType, e.ToRoute, e.PreviousRoute
	return p.client.Enqueue(ph.Capture{DistinctId: p.who(), Event: "$pageview", Properties: props, Timestamp: e.Timestamp})
}

func (p *Integration) Reset(ctx context.Context) error {
	if err := p.BaseIntegration.Reset(ctx); err != nil {
		return err
	}
	p.mu.Lock()
	p.distinctID = ""
	p.mu.Unlock()
	return nil
}

// Shutdown closes the client, which flushes pending messages.
func (p *Integration) Shutdown(context.Context) error {
	if p.client == nil {
		return nil
	}
	return p.client.Close()
}
